### ISSUE TITLE
`split/3` fails if one of the lists contains the delimiter element

### ISSUE THREAD

**Issue by jamesnvc** [mndrix/list_util#30]
I'm not sure if this is intentional, but removing the `dif(H, Div)` on line 73 seems to allow this and work okay?

**Comment by mndrix** [comment]
>
> I'm not sure if this is intentional, but removing the dif(H, Div) on line
> 73 seems to allow this and work okay?
>

Thanks for the report.  I know dif(H,Div) is intentional, but I don't
remember the details.  I should have some time to look into it early next
week.

Can you give me a concrete example of the split/3 call you're making and
how you expect it to behave?

**Comment by jamesnvc** [comment]
Sure; I'm trying something like this:

```
?- list_util:split(S, 0' , [`foo baz`, `bar`]).
```
Expected output: 
```
S = `foo baz bar`.
```

Actual output: `false`.

**Comment by mndrix** [comment]
I think your expectation is reasonable.  When called in the mode `split(-,+,+)`, there's no ambiguity of what the right answer should be.  If we support that, the question becomes: how should the predicate behave with similar input but with the inverse mode.  Namely, in `split("foo baz bar", 0' , Xs)` what should `Xs` be?

One answer is to iterate all possible splits like:

```
Xs = ["foo", "baz", "bar"];
Xs = ["foo", "baz bar"];
Xs = ["foo baz", "bar"];
Xs = ["foo baz bar"].
```

That might be correct, but it doesn't seem particularly useful. It also changes the predicate's declared determinism from `det` to `multi` in that mode.  It's also less efficient in the common case (wanting just the first result) since it leaves choicepoints or requires wrapping split/3 with once/1.

Maybe it's OK to add a special case to support `split(-,+,+)` as you expected and just document that that particular operation is not reversible if the separator is present inside a separated string.  That pollutes the semantics a little bit since `"foo baz"` not a possible output of `split/3`.

Another option would be to make a separate `join/3` predicate which only operates in a single mode but gives the behavior you wanted.

@jamesnvc Based on your use case, does any of these approaches seem more useful than the others?

**Comment by jamesnvc** [comment]
Ah, of course, that makes sense!

Hm...I guess adding `join/3` would be the most straightforward, to not break existing semantics, but maybe point out in the documentation for `split/3` that it'll fail if the list contains the delimiter (which I suppose means that it should be `semidet`)? 

 On the other hand, it might seem kind of confusing that there's a separate-but-very-similiar predicate; on that basis, I suppose special-casing `split(-,+,+)` would be the most likely to do what the user expects (and keeps it `det` (I think?)), at the cost of making the implementation a little bit ugly.

As a user of the library, I suppose I would prefer that second option of special-casing...

**Comment by mndrix** [comment]
> I suppose I would prefer that second option of special-casing

Me too.  It keeps the user-visible portion clean while essentially retaining the current semantics.  I'll work on implementing this in the next couple days.

**Comment by mndrix** [comment]
@jamesnvc Thanks for your help on this.  The implementation ended up being cleaner than it was before, so it seems to be a win all around.

I've released these changes as part of v0.13.0

**Comment by jamesnvc** [comment]
Excellent, I’m glad to hear it! Thank-you for your fast response & for making such a helpful library 😁
